MapboxVision/Services/VideoSource/ObservableVideoSource.swift

```diff
@@ -4,38 +4,67 @@ import Foundation
  Helper class handling observers: storing, releasing, notifying.
  Observers are held weakly by the instance of the class.
  You may inherit your video source from this class to avoid handling observers yourself.
+
+ - Important
+ A deadlock can occur when add/remove methods are called inside the notify's closure.
  */
 open class ObservableVideoSource: NSObject, VideoSource {
+    // MARK: - Open properties
+
     /// Provides default value of `isExternal` parameter.
     /// Override if your video source is represented by a module separate from the device.
     open var isExternal = true
 
+    // MARK: - Private properties
+
+    private var observations = [ObjectIdentifier: Observation]()
+    private var lock: UnsafeMutablePointer<os_unfair_lock>
```

Comment:
```suggestion
    private let lock = os_unfair_lock()
```
Then there's no need for `init` and `deinit`.
You can pass the lock as `os_unfair_lock_lock(&lock)`

Reply:
Unfortunately, we can't do this as it will potentially lead to the issue with the latest Swift.

According to `Swift memory management` (https://github.com/apple/swift/blob/master/docs/OwnershipManifesto.md):

```
We should add a core rule to the ownership system, called the Law of Exclusivity, which requires the implementation to prevent variables from being simultaneously accessed in conflicting ways. (For example, being passed inout to two different functions
```

More generally:

```
If a storage reference expression evaluates to a storage reference that is implemented by a variable, then the formal access duration of that access may not overlap the formal access duration of any other access to the same variable unless both accesses are reads.
```

Some details are here: https://forums.swift.org/t/exposing-the-memory-locations-of-class-instance-variables/30584

Reply:
I need to dive into this to get more context. Could you elaborate on what won't work in this particular use-case?